### ISSUE TITLE
[Bug]: test_logprobs.py::test_spec_decode_logprobs not using eager mode even though comment says it is

### ISSUE THREAD

**Issue by rasmith** [vllm-project/vllm#46657]
### Your current environment

I'm sorry, but this might reveal some internal information, so I cannot do this right now.

### 🐛 Describe the bug

The `test_logprobs.py::test_spec_decode_logprobs` test claims to use eager mode during the test but when I checked (by checking in llm.py) enforce_eager=False was encountered.

The relevant comment from the [source](https://github.com/vllm-project/vllm/blob/main/tests/v1/sample/test_logprobs.py#L1080)

```
   Runs the base model and spec decode model sequentially, ensuring
    only one LLM instance is alive at a time to avoid GPU memory
    contention. Both use identical chunked prefill settings and eager
    mode to control for infrastructure differences.
```
When running this test, I encountered:

```
FAILED v1/sample/test_logprobs.py::test_spec_decode_logprobs[ngram-raw_logprobs]
    Logprob mismatch: ref=-2.724879503250122 spec=-2.5109238624572754 diff=0.21395564079284668 (token=' ')

FAILED v1/sample/test_logprobs.py::test_spec_decode_logprobs[ngram-processed_logprobs]
     Logprob mismatch: ref=-2.724879503250122 spec=-2.5109238624572754 diff=0.21395564079284668 (token=' ')
```

The following change:

<code>
     llm_kwargs = dict(
        max_logprobs=5,
        max_model_len=max_model_len,
        seed=42,
        logprobs_mode=logprobs_mode,
        gpu_memory_utilization=0.4,
        # Force the same prefill chunking for both the base model and
        # spec decode model so the comparison isolates spec decode.
        enable_chunked_prefill=True,
        max_num_batched_tokens=32,
        enable_prefix_caching=False,
        <strong>enforce_eager=True</strong>,
        **GPU_DETERMINISM_KWARGS,
    )
</code>

Results in

`== 12 passed, 17 warnings in 434.99s (0:07:14) ==`

This test could fail if RNG changes, this could differ from platform to platform, or if the runtime changes (future versions of ROCm, etc.)

@AndreasKaratzas 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by AndreasKaratzas** [comment]
This seems highly underspecified issue. The only thing I can say is that the docstring is not accurate. Other than that the rest should be fine.

**Comment by AndreasKaratzas** [comment]
Is there a build kite link where this is failing?

**Comment by rasmith** [comment]
> Is there a build kite link where this is failing?

BK link: [https://buildkite.com/vllm/amd-ci/builds/9582/list?sid=019ecfa8-ea95-46a7-9889-bd209c7b1b5d&tab=output](https://buildkite.com/vllm/amd-ci/builds/9582/list?sid=019ecfa8-ea95-46a7-9889-bd209c7b1b5d&tab=output)

Changes in RNG seem to be affecting this.  Fails only with `TRITON_ATTN `backend (works with `ROCM_ATTN`), and `ngram `speculative decode method.  I checked all handwritten Triton kernels and all inductor generated fused Triton kernels that were called (since `enforce_eager=False`), and compared against most recent `vllm/vllm-openai-rocm:v0.23.0` vs most recent `TheRock` docker and all kernels evaluated the same (same outputs for same inputs).  

I checked in both containers and there are differences between generated values (`torch.randn`  -> `torch.save` then `torch.testing.assert_close`)

Setting` enforce_eager=True` might make this test less flaky, especially as RNG could change in the future.

**Comment by mawong-amd** [comment]
I'm not convinced just yet that this issue is correctly attributed to eager mode. Since this test is flaky, the fact that it passes some number of times with eager mode enabled could just be a fluke.
The comment mentions eager mode, sure, but there's no reason to believe that the comment is correct. The RNG seed is also fixed, which ought to rule that out as a factor, and `max_num_seqs` is indeed set to 1, so batch invariance likely isn't implicated here.

More investigation is needed before more can be said about this issue. If enforce eager is indeed implicated, there would need to be a concrete explanation pointing to different code paths with different numerical behavior under enforce eager (so in other words, due to either CUDA graphs or torch.compile or both).

**Comment by rasmith** [comment]
@AndreasKaratzas  @mawong-amd  How much is this failing in main CI? It's reliably failing when running with `rocm/vllm-dev:nightly-therock` (3.12 Python).

**Comment by mawong-amd** [comment]
Is this test ever passing on that container?

If it is, then this is flaky and there should be more investigation as to what is non-deterministic in this test, so we can control for it.
To give a hypothetical example, if it turns out that updated BLAS libraries in that container call into non-deterministic (because of atomic add) splitK GEMMs, then the fix wouldn't be to turn on `enforce_eager` here, but to ensure that the logprobs are still sensible even if not bitwise-exact.

If it isn't, then some effort should be put into understanding why the different codepaths lead to deterministically different results.
For instance, it may turn out that the CUDA graph capture calls into a different kernel at some point due to padding in some dimension, and this kernel has a different accumulation order which leads to numerical divergence.

Point is, turning on `enforce_eager` is a reasonably large change from a code path perspective: it turns off both `torch.compile` and also CUDA graphs, both of which are performance features that users rely on by default. If possible, we ought to investigate more to see if we can more narrowly scope this change.
For a start, we might use `compilation_config={"cudagraph_mode": CUDAGraphMode.NONE}` to isolate the change to CUDA graphs rather than `torch.compile`, and similarly we might also test `torch.compile` while leaving CUDA graphs on.

**Comment by jyan-R** [comment]
We hit `test_spec_decode_logprobs` failing while running vLLM's NVIDIA CI on an A100, and found this thread while tracking it down. We aren't working on ROCm — everything below is from NVIDIA hardware, on unpatched upstream `a3561ef8e`, with no changes to the test's requests, tolerances, assertions, or execution mode.

**This is not a reproduction of the ROCm report above.** Our failure signature is the `decoded_token` assertion rather than the score tolerance, and we are not claiming a shared root cause. But several questions raised in this thread now have measured answers.

### The CUDA graphs vs `torch.compile` split

The experiment suggested above — `compilation_config={"cudagraph_mode": CUDAGraphMode.NONE}` — does separate them. Full matrix, `ngram` x `raw_logprobs` x `top_logprobs=3`, each cell a fresh process:

| config | A100 80GB PCIe | H100 PCIe |
|---|---|---|
| upstream default | FAIL | PASS |
| `max_num_seqs=1` | FAIL | PASS |
| `attention_backend=TRITON_ATTN` | FAIL | PASS |
| `max_num_seqs=1` + `TRITON_ATTN` | **PASS** | PASS |
| `compilation_config={"cudagraph_mode": "NONE"}` | FAIL | PASS |
| `enforce_eager=True` | **PASS** | PASS |
| upstream default, `top_logprobs=5` | FAIL | **FAIL** |

Turning off CUDA graphs alone is not enough; only `enforce_eager=True` passes. But `enforce_eager` is not special — `max_num_seqs=1 + TRITON_ATTN` also passes, while neither of those knobs passes on its own. `max_num_seqs=1` alone actually makes it worse, raising the mismatch count from 1 to 2.

The observation earlier in this thread that it fails with `TRITON_ATTN` but works with `ROCM_ATTN` has the same shape on CUDA: swapping the attention backend flips the verdict, but no backend is the "correct" one. That is what you would expect if these knobs move the top-k cut rather than repair a computation — which is also consistent with the finding reported here that every hand-written and inductor-generated Triton kernel produced identical output for identical input.

The last row also shows this isn't Ampere-specific: the same H100 fails at `top_logprobs=5`.

### `enforce_eager` passing does not mean it fixes anything

The scepticism above about it possibly being a fluke was well placed. We captured the full pre-topk logits tensor and compared it bit-for-bit between the two engines:

| config (A100, `ngram` x `raw_logprobs`) | test | vocab entries differing | max \|Δ\| | mean \|Δ\| |
|---|---|---|---|---|
| upstream default | FAIL | 64.69% | 0.125 | 1.27e-02 |
| `cudagraph_mode=NONE` | FAIL | 66.08% | 0.125 | 1.36e-02 |
| `max_num_seqs=1` | FAIL | 77.23% | 0.125 | 1.54e-02 |
| `enforce_eager=True` | **PASS** | 68.44% | **0.1875** | **1.68e-02** |

**The configuration that passes has the largest max and mean divergence of the four; the upstream default, which fails, has the smallest.** Disabling `torch.compile` does not reduce the disagreement between the two paths at all.

What it does instead: in all 8 of our `decoded_token` failures, the two tokens being compared have **bit-identical raw BF16 logits in the spec engine**. When two candidates are tied, which one lands at rank k comes down to `torch.topk`'s unspecified tie-breaking. Ties at the cut are not rare here either — 1 to 4 of the 20 generated positions per configuration, in both engines, with the largest tie group containing 3 tokens. So `enforce_eager` passes because it happens to land the cut off a tie, not because it made anything more correct.

### "Sensible even if not bitwise-exact" — they are

Against the criterion raised above, across the whole 128k vocabulary: at most 1 BF16 code step within the top-8, 1–2 within the top-64, 3 within the top-1024, and an absolute difference of ≤ 0.1875 against a logit range of roughly −6 to +23. The two paths agree to close to the BF16 representation limit.

Also worth noting: **in every configuration, including all the failing ones, the sampled output tokens from the two engines are identical.**

### What we have not established

We have **not** located the source of the numerical difference — which kernel, which fusion, which accumulation order — and we did not run a layer-by-layer probe. We also cannot say whether the ROCm score-tolerance failure reported here shares a mechanism with what we see; the failure message only prints the ref-side token, so a mis-paired comparison and a genuine numerical deviation look identical in the log.

### On the A100 / ROCm connection

We don't claim our A100 results transfer to MI300. But one thing might be worth a look: ROCm CI pins exactly `max_num_seqs=1` + `TRITON_ATTN`, and on A100 that is the one combination that passes, with neither knob sufficient alone. If someone with an MI300 is able to run this at `top_logprobs=5` and check whether there are exact ties at the cut, that would settle whether the two platforms are seeing the same thing.

We also filed #53620 about a separate CI-side problem that keeps these 12 nodes from running on NVIDIA runners at all.

*AI assistance was used for analysis and scripting; all experiments were run on real GPUs and every number was checked against the raw logs.*

**Comment by rasmith** [comment]
affected test no longer failing in Rock 7.14 CI
